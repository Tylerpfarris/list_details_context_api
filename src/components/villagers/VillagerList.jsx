import React from 'react';
import { useVillagersFetch } from '../hooks/useVillagersFetch';
import Villager from './Villager';
import { Link } from 'react-router-dom';
import { Container, Button, makeStyles, Typography } from '@material-ui/core';

const useStyles = makeStyles((theme) => ({
  themeDark: {
    backgroundColor: 'black',
    color: 'white',
    border: 'solid 1px white',
    margin: '5px',
  },
  themeLight: {
    backgroundColor: 'white',
    color: 'black',
    border: 'solid 1px black',
    margin: '5px',
  },
}));
function VillagerList() {
  const classes = useStyles();
  const { villagers, loading, page, setPage, darkMode } = useVillagersFetch();
  const pagedVillagersList = villagers.slice((page - 1) * 24, page * 24);

  const ePageDecrement = () => {
    setPage((prev) => prev - 1);
  };
  const ePageIncrement = () => {
    setPage((prev) => prev + 1);
  };

  if (loading) return <h1 data-testid="loading">Loading...</h1>;
  else return (
    <Container>
      <Container
        style={{
          marginTop: '20px',
          display: 'flex',
          flexDirection: 'row',
          alignItems: 'center',
          justifyContent: 'center',
        }}
      >
        {page === 1 ? (
          <Button
            data-testid="back-button-disabled"
            className={darkMode ? classes.themeDark : classes.themeLight}
            disabled
            style={{
              color: 'grey',
            }}
            onClick={ePageDecrement}
          >
            <Typography>back</Typography>
          </Button>
        ) : (
          <Button
            data-testid="back-button"
            className={darkMode ? classes.themeDark : classes.themeLight}
            onClick={ePageDecrement}
          >
            <Typography>back</Typography>
          </Button>
        )}
        <Typography data-testid="page-counter">
          {page} / {Math.ceil(villagers.length / 24)}
        </Typography>

        {page === Math.ceil(villagers.length / 24) ? (
          <Button
            data-testid="next-button-disabled"
            className={darkMode ? classes.themeDark : classes.themeLight}
            disabled
            style={{
              color: 'grey',
            }}
            onClick={ePageIncrement}
          >
            <Typography>next</Typography>
          </Button>
        ) : (
          <Button
            data-testid="next-button"
            className={darkMode ? classes.themeDark : classes.themeLight}
            onClick={ePageIncrement}
          >
            <Typography>next</Typography>
          </Button>
        )}
      </Container>
      <ul
        aria-label="villager-list"
        style={{
          display: 'flex',
          flexWrap: 'wrap',
          flexDirection: 'row',
          listStyle: 'none',
        }}
      >
        {pagedVillagersList.map((villager) => (
          <li
            key={villager.villagerId}
            style={{
              margin: '10px',
            }}
          >
            <Link
              style={{ textDecoration: 'none', textAlign: 'center' }}
              to={`details/${villager.villagerId}`}
            >
              <Villager {...villager} />
            </Link>
          </li>
        ))}
      </ul>
  </Container>
    )
}

export default VillagerList;
