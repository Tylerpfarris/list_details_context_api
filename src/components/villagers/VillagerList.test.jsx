/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, screen, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';
import { rest } from 'msw';
import { setupServer } from 'msw/node';

import animalCrossingCharacterList from '../../animalCrossingCharacterList.json';
import VillagerList from './VillagerList';
import { ThemeProvider } from '../state/ThemeContext';

const server = setupServer(
  rest.get(`https://acnhapi.com/v1a/villagers`, (req, res, ctx) => {
    return res(ctx.json(animalCrossingCharacterList));
  })
);

describe('VillagerList', () => {
  beforeAll(() => server.listen());
  afterAll(() => server.close());
  afterEach(cleanup);

  it('renders a list of animal crossing characters on the screen', async () => {
    render(
      <ThemeProvider>
        <MemoryRouter>
          <VillagerList />
        </MemoryRouter>
      </ThemeProvider>
    );

    const disabledBackButton = screen.getByTestId('back-button-disabled');
    expect(disabledBackButton).toHaveTextContent('back');

    const nextButton = screen.getByTestId('next-button');
     expect(nextButton).toHaveTextContent('next');

    userEvent.click(nextButton);

    const backButton = screen.getByTestId('back-button');
     expect(backButton).toHaveTextContent('back');

    userEvent.click(nextButton);

    const listEl = await screen.getByRole('list', 'villager-list');
    expect(listEl).toMatchSnapshot();

    // waitFor(() => {
    //   const disabledNextButton =  screen.getByTestId('next-button-disabled');
    //   expect(disabledNextButton).toHaveTextContent('next')

    // })
  });
});
