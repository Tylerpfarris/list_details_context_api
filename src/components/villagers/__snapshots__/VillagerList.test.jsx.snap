// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VillagerList renders a list of animal crossing characters on the screen 1`] = `
<ul
  aria-label="villager-list"
  style="display: flex; flex-wrap: wrap; flex-direction: row; list-style: none;"
/>
`;
